### PR TITLE
[Clang][DebugInfo] Clang generates an extra spurious unnamed 'dbg.declare'

### PR DESCRIPTION
**Pull request by CarlosAlbertoEnciso**
Do not emit call to llvm.dbg.declare when the variable declaration
is a DecompositionDecl as its instance class is always unnamed.

The emitted debug declare looks like:
```
call void @llvm.dbg.declare(metadata ..., metadata !xx, metadata ...)
!xx = !DILocalVariable(scope: !..., file: !..., line: ..., type: !...)
```

### CODE REVIEW

File: clang/test/CodeGenCXX/debug-info-structured-binding-field.cpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  int Fld_2;
+};
+__attribute__((optnone)) Tuple get() { return {10, 20}; }
+
+// CHECK-LABEL: define dso_local noundef i32 @main
+// CHECK:      %retval = alloca i32, align 4
+// CHECK-NEXT: [[T0:%.*]] = alloca %struct.Tuple, align 4
+// CHECK:      call void @llvm.dbg.declare(metadata ptr [[T0]], metadata {{.*}}, metadata !DIExpression())
+// CHECK:      call void @llvm.dbg.declare(metadata ptr [[T0]], metadata {{.*}}, metadata !DIExpression(DW_OP_plus_uconst, {{[0-9]+}}))
+// CHECK-NOT:  call void @llvm.dbg.declare(metadata ptr [[T0]], metadata {{.*}}, metadata !DIExpression())
```
Comment (OCHyams) [inline clang/test/CodeGenCXX/debug-info-structured-binding-field.cpp:14]: It doesn't look like this test covers anything that `clang/test/CodeGenCXX/debug-info-structured-binding.cpp` doesn't already?

IMO no need to add this one. Please can you modify the other test to check that the DILocalVariables in the dbg.declares have the expected names (i.e., confirm there's no unnamed variable).

File: clang/test/CodeGenCXX/debug-info-structured-binding.cpp [hunk 2]
```diff
@@ -2,10 +2,8 @@
```
Comment (OCHyams) [inline clang/test/CodeGenCXX/debug-info-structured-binding.cpp:2]: Please can you add `--implicit-check-not="call void @llvm.dbg.declare"` to the `FileCheck` command?

**Review by OCHyams — COMMENTED** [review]
This SGTM.

I don't think the unnamed variable serves any useful purpose but may not being imaginative enough. I can't see any discussion on the topic when the code was added in [D119178](https://reviews.llvm.org/D119178) . CC the author, @shafik (or @adrian-prantl) - is there a use-case for the anonymous variable that covers the whole of the structured binding storage?

**Review by adrian-prantl — APPROVED** [review]
Aside from other open comments the spirit of the change LGTM.

File: clang/test/CodeGenCXX/debug-info-structured-binding.cpp (hunk 2 above)
Comment (CarlosAlbertoEnciso) [reply clang/test/CodeGenCXX/debug-info-structured-binding.cpp:12]: Added the `--implicit-check-not` option.

**Review by OCHyams — APPROVED** [review]
New changes LGTM, thanks!

**Comment by CarlosAlbertoEnciso** [comment]
@adrian-prantl, @OCHyams Thanks for your reviews.

**Pull request merged**